Updating the descriptions for the escape/encode functions

julieheard opened the pull request:
Adding explanations and examples to the encode and escape functions.
This is a documentation only change.

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done
I have manually tested the output of these functions multiple times in Jelly and in java and put the output in the documentation. 

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries
- Added descriptions to the escape and encode functions in the hudson.Functions library


### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7606"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

julieheard committed:
Updating the descriptions for the escape/encode functions

Adding explanations and examples to the encode and escape functions

julieheard committed:
Updated description for encode 

I have edited some of the descriptions to bring them in line with the documentation over on husdon.Util

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -759,9 +759,12 @@ public static boolean isMultiline(String s) {
 
     /**
-     * Encodes a string into UTF-8 encoding
+     * Partially encodes a string into UTF-8 encoding. It uses the hudson.Util encode method to escape non-ASCII characters in URL.
```
> hudson.Util encode

Irrelevant implementation detail?

Also, the output example has leading and trailing spaces not apparent in the input string.

julieheard replied:
Hi, thank you for reviewing this PR.

I wanted to convey that this doesn't fully encode the whole string as UTF-8, it only encodes some of the characters. I am open to how better to word this so that it doesn't just read as 'encodes the whole string into utf-8'

Thank you for spotting the spaces in the example, I have done a commiit to change that.

julieheard committed:
Changed leading and trailing spaces on encode

I left one space at the end and added a note to say how a blank space is rendered

julieheard committed:
Removed trailing whitespace to satisfy checkstyle

julieheard committed:
Merge branch 'master' into master

julieheard commented:
Looks like the build failure is due to it not liking certain chars in my examples. It doesn't like these chars: **& > <** 
Do you know a sneaky way of getting roung this so we can keep the examples, or shall I write more _'Note: Less than symbol will encode to &lts;'_ bits?

NotMyFault committed:
Preserve formatting and escape HTML tags

NotMyFault commented:
I proposed a change that preserves all formatting and escapes html tags. That should get us a green build 🤞🏻

NotMyFault reviewed (approved):
Looks good so far :)

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -758,6 +758,19 @@ public static boolean isMultiline(String s) {
+    /**
+     * Partially encodes a string into UTF-8 encoding. It uses the {@link hudson.Util} encode method to escape non-ASCII characters in URL.
+     * <pre>
+     * Input example  1: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
+     * Output example 1: !"%C2%A3$%^&amp;*()_+}{:@~?&gt;&lt;|%C2%AC`,./;'#[]-=%20
```
I'd put the space somewhere in the middle. As a trailing space it'll just keep getting erased.
```suggestion
     * Input example  1: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]- =
     * Output example 1: !"%C2%A3$%^&amp;*()_+}{:@~?&gt;&lt;|%C2%AC`,./;'#[]-%20=
```

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -766,6 +779,13 @@ public static String encode(String s) {
+     * Input example  1: &amp; " ' &lt; &gt;
+     * Output example 1: %26+%22+%27+%3C+%3E
+     * Input example  2: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
+     * Output example 2: %21%22%C2%A3%24%25%5E%26*%28%29_%2B%7D%7B%3A%40%7E%3F%3E%3C%7C%C2%AC%60%2C.%2F%3B%27%23%5B%5D-%3D++
```
Also confusing trailing spaces out of nowhere.

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +796,33 @@ public static String urlEncode(String s) {
+    /**
+     * Escapes HTML unsafe characters
+     * <pre>
+     * Input example  1: &amp; " ' &lt; &gt;
+     * Output example 1: &amp; &quot; &#039; &lt; &gt;
+     * Input example  2: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
+     * Output example 2: !&quot;£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;&#039;#[]-=&nbsp;
```
Outputs need to be double-escaped, otherwise they're going to be rendered the same as input in HTML Javadoc. From IntelliJ preview:

<img width="598" alt="Screenshot" src="https://user-images.githubusercontent.com/1831569/215861445-716b5900-0731-4232-92a5-be449c42733c.png">

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +796,33 @@ public static String urlEncode(String s) {
+     * Notes:
+     * <ul>
+     * <li>2 consecutive blank spaces will render as &nbsp; (which is a non-breaking space char)</li>
+     * <li>This method will render an apostrophe as &#039;</li>
```
as above

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +796,33 @@ public static String urlEncode(String s) {
+    /**
+     * Escapes XML unsafe characters
+     * <pre>
+     * Input example  1: &lt; &gt; &amp;
+     * Output example 1: &lt; &gt; &amp;
+     * Input example  2: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
+     * Output example 2: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
```
Based on these examples, this doesn't do anything?

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -788,6 +831,16 @@ public static String xmlUnescape(String s) {
+    /**
+     * Escapes HTML unsafe characters
+     * <pre>
+     * Input example  1: &amp; " ' &lt; &gt;
+     * Output example 1: &amp; &quot; &#39; &lt; &gt;
+     * Input example  2: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
+     * Output example 2: !&quot;£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;&#39;#[]-=
```
Same problem as above, these will render identically in HTML Javadoc output.

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +796,33 @@ public static String urlEncode(String s) {
+     * </pre>
+     * Notes:
+     * <ul>
+     * <li>2 consecutive blank spaces will render as &nbsp; (which is a non-breaking space char)</li>
```
It goes further than this: A newline character will render as `<br>` HTML tag.

The purpose of this method (originally) to faithfully render a human-entered description. So newlines translate to line break tags, consecutive spaces are retained as nbsp, and we escape HTML special chars to not get markup.

This should be explained, and `@see #xmlEscape` be added. Most use cases are probably better suited by the other method.

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard committed:
Changed escaping on examples so they rendered properly in the javadoc

I was tempted to use {@literal &amp;} notation but this added blank spaces, making the examples look strange.  Instead I have used the &amp;amp; notation.

julieheard commented:
![image](https://user-images.githubusercontent.com/55280278/216307388-8918d2ac-424c-4664-9700-53861a550a23.png)
![image](https://user-images.githubusercontent.com/55280278/216307608-2c5ab3f3-e486-46b6-ac54-ee62c7c7caca.png)
![image](https://user-images.githubusercontent.com/55280278/216310023-456e2b4c-92b6-4289-b131-a60cf41621d5.png)
![image](https://user-images.githubusercontent.com/55280278/216308803-ab05aef2-7dcd-42f8-b079-52c6753bed2c.png)
![image](https://user-images.githubusercontent.com/55280278/216307884-538597d4-534f-4510-9325-f18ba7eb01c9.png)

julieheard committed:
Merge branch 'jenkinsci:master' into master

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +797,40 @@ public static String urlEncode(String s) {
 
+    /**
+     * Escapes HTML unsafe characters.
+     *<br/>
```
```
error: self-closing element not allowed
      *<br/>
       ^

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -758,6 +758,20 @@ public static boolean isMultiline(String s) {
     }
 
+    /**
+     * Partially encodes a string into UTF-8 encoding. It uses {@link hudson.Util#encode} to escape
+     * non-ASCII characters in URL.
```
Partial seems unnecessarily vague, and mentioning only UTF-8 seems misleading, given this percent-encodes the string (which are orthogonal).

How about:

```suggestion
     * Percent-encodes space and non-ASCII UTF-8 characters for use in URLs.
```

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +797,40 @@ public static String urlEncode(String s) {
     }
 
+    /**
+     * Escapes HTML unsafe characters.
+     *<br/>
+     * The purpose of this method (originally) was to faithfully render a human-entered description.
+     * So newlines translate to line break tags, consecutive spaces are retained as nbsp, and we escape HTML special chars to not get markup.
```
```suggestion
     * Transforms the input string so it renders as written in HTML output: newlines are converted to HTML line breaks, consecutive spaces are retained as {@code &amp;nbsp;}, and HTML metacharacters are escaped.
```

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -776,10 +797,40 @@ public static String urlEncode(String s) {
+     * Input example  2: !"£$%^&amp;*()_+}{:@~?&gt;&lt;|¬`,./;'#[]-=
+     * Output example 2: !&amp;quot;£$%^&amp;amp;*()_+}{:@~?&amp;gt;&amp;lt;|¬`,./;&amp;#039;#[]-=
+     * </pre>
+     * Notes:
+     * <ul>
+     * <li>2 consecutive blank spaces will render as &amp;nbsp; (which is a non-breaking space char)</li>
+     * <li>This method will render an apostrophe as &amp;#039;</li>
+     * <li>A newline character \n will render as &lt;br&gt; HTML tag</li>
+     * </ul>
```
```suggestion
```

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -788,6 +839,16 @@ public static String xmlUnescape(String s) {
     }
 
+    /**
+     * Escapes HTML unsafe characters
```
```suggestion
     * Escapes a string so it can be used in an HTML attribute value.
```

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1569,6 +1630,13 @@ public static <T> Collection<T> emptyList() {
     }
 
+    /**
+     * Escapes Javascript unsafe characters
```
```suggestion
     * Escape a string so variable values can be used in inline JavaScript in views.
     * Note that inline JavaScript and especially passing variables is discouraged, see the documentation for alternatives.
```

daniel-beck commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1569,6 +1630,13 @@ public static <T> Collection<T> emptyList() {
+     * <pre>
+     * Input example : \ \\ ' "
+     * Output example: \\ \\\\ \' \"
+     * </pre>
```
```suggestion
     * </pre>
     * @see <a href="https://www.jenkins.io/doc/developer/security/xss-prevention/#passing-values-to-javascript">Passing values to JavaScript</a>
```

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard committed:
Update core/src/main/java/hudson/Functions.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

julieheard commented:
Agreed with all changes, thank you so much @daniel-beck

timja commented:
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
